src/dbnode/storage/shard.go

```diff
@@ -2660,74 +2661,217 @@ func (s *dbShard) Repair(
 
 func (s *dbShard) AggregateTiles(
 	ctx context.Context,
-	reader fs.DataFileSetReader,
 	sourceNsID ident.ID,
-	sourceBlockStart time.Time,
-	sourceShard databaseShard,
+	sourceShardID uint32,
+	blockReaders []fs.DataFileSetReader,
+	sourceBlockVolumes []shardBlockVolume,
 	opts AggregateTilesOptions,
-	wOpts series.WriteOptions,
+	targetSchemaDescr namespace.SchemaDescr,
 ) (int64, error) {
-	latestSourceVolume, err := sourceShard.latestVolume(sourceBlockStart)
+	blockReadersToClose := make([]fs.DataFileSetReader, 0, len(blockReaders))
+	defer func() {
+		for _, reader := range blockReadersToClose {
+			reader.Close()
+		}
+	}()
+
+	maxEntries := 0
+	for sourceBlockPos, blockReader := range blockReaders {
+		sourceBlockVolume := sourceBlockVolumes[sourceBlockPos]
```

Comment:
As discussed let's add a length check to make sure `len(blockReaders) == len(sourceBlockVolumes)` and return error if no match, so that a code bug doesn't cause a panic in future.

Reply:
👍